java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/consumer/ArrayConsumer.java

```diff
@@ -90,13 +97,12 @@ public void consume(ResultSet resultSet) throws SQLException, IOException {
         int count = 0;
         try (ResultSet rs = array.getResultSet()) {
           while (rs.next()) {
-            ensureInnerVectorCapacity(innerVectorIndex + count + 1);
```

Comment:
It's because of how a ListVector is laid out in memory. The list `[[1, 2], [], [3, 4, 5]]` is represented as the child vector `[1, 2, 3, 4, 5]` and the offsets `[0, 2, 2, 5]`. `ensureInnerVectorCapacity` is resizing the child vector, so when we call `consume` for the last element, we want to ensure the child vector has enough capacity for the current elements, along with all the previous elements, and it looks like that's what `innerVectorIndex` is tracking. 

In other words when we call `consume` for what will be `[3, 4, 5]` we need to ensure the child vector has space for at least 3, 4, 5, … elements not 1, 2, 3… elements.

Reply:
I'm still not entirely convinced we can remove `innerVectorIndex` here?

Reply:
I've reverted the change to `innerVectorIndex`. Only thing I've done is reset it to 0 when we reset inner vector.

I've added unit tests for Arrays which covers the two issues when reusing `VectorSchemaRoot`:
1) NPE error from `ArrowVectorIterator`.
2) `ArrayConsumer` not reseting the delegate consumer.